Tests/MongoSwiftTests/SpecTestRunner/Match.swift

```diff
@@ -10,34 +10,58 @@ internal protocol Matchable {
     /// This assumes `expected` is NOT a placeholder value (i.e. 42/"42"). Use `matches` if `expected` may be a
     /// placeholder.
     /// https://github.com/mongodb/specifications/tree/master/source/connection-monitoring-and-pooling/tests#spec-test-match-function
-    func contentMatches(expected: Any) -> Bool
+    func contentMatches(expected: Self) -> Bool
+
+    /// Determines if this value is considered a wildcard for the purposes of the MATCHES function.
+    func isPlaceholder() -> Bool
 }
 // swiftlint:enable line_length
 
 extension Matchable {
+    internal func isPlaceholder() -> Bool {
+        return false
+    }
+
     /// Returns whether this MATCHES the expected value according to the function defined in the spec.
-    internal func matches(expected: Any) -> Bool {
-        return isPlaceholder(expected) || self.contentMatches(expected: expected)
+    internal func matches<T: Matchable>(expected: T) -> Bool {
+        guard !expected.isPlaceholder() else {
+            return true
+        }
+
+        guard let expected = expected as? Self else {
+            return false
+        }
+        return self.contentMatches(expected: expected)
+    }
+}
+
+extension Matchable where Self: Equatable {
+    internal func contentMatches(expected: Self) -> Bool {
+        return self == expected
+    }
+}
+
+extension Int: Matchable {
+    internal func isPlaceholder() -> Bool {
+        return self == 42
+    }
+}
+
+extension String: Matchable {
+    internal func isPlaceholder() -> Bool {
+        return self == "42"
     }
 }
 
 /// Extension that adds MATCHES functionality to `Array`.
-extension Array: Matchable {
-    internal func contentMatches(expected: Any) -> Bool {
-        guard let expected = expected as? [Any], expected.count <= self.count else {
+extension Array: Matchable where Element: Matchable {
+    internal func contentMatches(expected: [Element]) -> Bool {
+        guard expected.count <= self.count else {
             return false
         }
 
         for (aV, eV) in zip(self, expected) {
```

Comment:
if you feel like it you could one line this with `allSatisfy`

Reply:
yep good point. updated.